socksio/singledispatchmethod.py

```diff
@@ -0,0 +1,42 @@
+"""Backport of @functools.singledispatchmethod to Python <3.7.
+
+Adapted from https://github.com/ikalnytskyi/singledispatchmethod
+removing 2.7 specific code.
+"""
+
+import functools
+import sys
+
+
+if sys.version_info >= (3, 8):
```

Comment:
Can we do feature detection instead of version detection here? ie `hasattr(functools, "singledispatchmethod")`